File: topic wise question/backtraking/js/1-permute_recur.js
function permute(ip, op, ans) {
  if (ip.length == 0) {
    ans.push(op);
    return;
  }

  let mpp = new Set();

  for (let i = 0; i < ip.length; i++) {
    if (!mpp.has(ip[i])) {
      mpp.add(ip[i]);
      let newIp = ip.substring(0, i) + ip.substring(i + 1);
      let newOp = op + ip[i];
      permute(newIp, newOp, ans);
    }
  }
}

function findPermute(s) {
  let ans = [];
  permute(s, "", ans);
  ans.sort();
  return ans;
}

const readline = require("readline");

const r1 = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

r1.question("Enter a string : ", function (s) {
  let permutation = findPermute(s);
  console.log(permutation.join(" "));
  r1.close();
});

File: topic wise question/backtraking/1-permute_recur.cpp
#include<bits/stdc++.h>
using namespace std;


//using recursion


void permute(string ip,string op,vector<string>&ans){
	if(ip.length()==0){
		ans.push_back(op);
		return;
	}

	unordered_set<char>mpp;
	for(int i = 0 ; i<ip.length();i++){
		if(mpp.find(ip[i])==mpp.end()){
			mpp.insert(ip[i]);
			string newIp = ip.substr(0,i)+ip.substr(i+1);
			string newOp = op+ip[i];
			permute(newIp,newOp,ans);
		}
	}
	
}

vector<string>findPermutation(string &s){
	vector<string>ans;
	permute(s,"",ans);
	return ans;
}

int main(){
	string s;
	cin>>s;
	
	vector<string>ans = findPermutation(s);
	sort(ans.begin(),ans.end());
	for(auto i:ans){
		cout<<i<<" "; 	
	}

	cout<<"\n";	
	return 0;
}